feat: add support for 'projects.getAncestry' API

rvandegrift opened the pull request:
The current `Project` implementation includes the immediate `parent` but provides no way to get the full ancestry.  This adds a property to expose that data.

rvandegrift committed:
feat: Add support for projects.getAncestry

tseaver commented on google/cloud/resource_manager/project.py:
```diff
@@ -102,6 +102,20 @@ def path(self):
+        https://cloud.google.com/resource-manager/reference/rest/v1beta1/projects/getAncestry
+        """
+        client = self._require_client(self._client)
+        resp = client._connection.api_request(
+            method="POST",
+            path='%s:getAncestry?alt=json' % (self.path),
+        )
```
JSON is the default response type for `getAncestry`.
```suggestion
        resp = client._connection.api_request(
            method="POST", path='%s:getAncestry' % (self.path),
        )
```

tseaver commented on tests/unit/test_project.py:
```diff
@@ -304,6 +304,23 @@ def test_undelete_with_reload_data(self):
         expected_get_request = {"method": "GET", "path": project.path}
         self.assertEqual(get_request, expected_get_request)
 
+    def test_fetch_project_ancestry(self):
+        PROJECT_ID = 'project-id'
+        GET_ANCESTRY_RESPONSE = {
+            'ancestor': [
+                {'resourceId': {'type': 'project', 'id': PROJECT_ID}},
+                {'resourceId': {'type': 'folder', 'id': '012345678901'}},
+                {'resourceId': {'type': 'folder', 'id': '987654321098'}},
+                {'resourceId': {'type': 'organization', 'id': '555555555555'}},
+            ],
+        }
+
+        connection = _Connection(GET_ANCESTRY_RESPONSE)
+        client = _Client(connection=connection)
+        project = self._make_one(PROJECT_ID, client)
+        ancestry = project.ancestry
+        self.assertEqual(ancestry, GET_ANCESTRY_RESPONSE['ancestor'])
+
```
We also need a test for the case that the response is missing the `ancestor` key:
```suggestion
    def test_fetch_project_ancestry_hit(self):
        PROJECT_ID = 'project-id'
        GET_ANCESTRY_RESPONSE = {}

        connection = _Connection(GET_ANCESTRY_RESPONSE)
        client = _Client(connection=connection)
        project = self._make_one(PROJECT_ID, client)

        self.assertEqual(project.ancestry, [])
        
    def test_fetch_project_ancestry_hit(self):
        PROJECT_ID = 'project-id'
        ANCESTORS = [
                {'resourceId': {'type': 'project', 'id': PROJECT_ID}},
                {'resourceId': {'type': 'folder', 'id': '012345678901'}},
                {'resourceId': {'type': 'folder', 'id': '987654321098'}},
                {'resourceId': {'type': 'organization', 'id': '555555555555'}},
            ]
        GET_ANCESTRY_RESPONSE = {'ancestor': ANCESTORS}
        connection = _Connection(GET_ANCESTRY_RESPONSE)
        client = _Client(connection=connection)
        project = self._make_one(PROJECT_ID, client)

        self.assertEqual(project.ancestry, ANCESTORS)
```

tseaver reviewed (changes requested):
Thanks for the patch!  In addition to my suggestions inline, I'm also ambivalent about having this as a property, rather than a method:  I would be inclined to spell it as `project.list_ancestors()` (we don't use properties to wrap other API-invoking methods).

Making it a method would make documenting the return value clearer, as well.

rvandegrift committed:
Respell ancestors property to list_ancestors()

rvandegrift committed:
JSON is the default response type for getAncestry

Co-authored-by: Tres Seaver <tseaver@palladion.com>

rvandegrift committed:
Add test case for missing ancestry in response

Co-authored-by: Tres Seaver <tseaver@palladion.com>

rvandegrift commented:
@googlebot I fixed it.

rvandegrift commented:
@tseaver I've implemented your suggestions and rebased on master.  googlebot seems confused over my CLA status, I've confirmed the commits are all from my CLA-covered email.

parthea commented:
CLA looks good for @rvandegrift. @tseaver, please check CLA for your username.

rvandegrift commented:
I don't need this change anymore, we've opted to stick with the discovery api clients in most places to avoid these kinds of feature gaps.
